File: src/tournament.ts

export enum State {
	Idle,
	Waiting,
	Shooting,
}

export type Light = 'red' | 'yellow' | 'green';
export type Shooters = 'A/B' | 'C/D';
export type Honks = 0 | 1 | 2 | 3;

export interface ITournamentState {
	currentLight: Light;
	currentShooters: Shooters;
	endsShot: number;
	totalEnds: number;
	secondsLeft: number;
	shooting: boolean;
	honks: Honks;
}

export interface ITournamentConfig {
	waitTimeInSeconds: number;
	arrowsPerEnd: number;
	secondsPerArrow: number;
	yellowPhaseInSeconds: number;
	numberOfEnds: number;
	alternatingShooters: boolean;
}

export class Tournament {
	private state: State = State.Idle;
	private endStart: Date = null;

	private currentLight: Light = 'red';
	private currentShooters: Shooters = 'A/B';
	private endsShot: 0;

	private configuration: ITournamentConfig = {
		alternatingShooters: true,
		arrowsPerEnd: 3,
		numberOfEnds: 20,
		secondsPerArrow: 40,
		waitTimeInSeconds: 10,
		yellowPhaseInSeconds: 30,
	};

	constructor() {
		this.initializeState();
	}

	public configure(configuration: ITournamentConfig): void {
		if (this.state === State.Idle) {
			this.configuration = configuration;
			this.initializeState();
		} else {
			throw new Error(`Invalid state "${this.state}" to change configuration`);
		}
	}

	public startEnd(date?: Date): void {
		this.endStart = new Date((date || new Date()).valueOf());
	}

	public endCurrentEnd(now?: Date): void {
		now = now || new Date();

		if (this.state !== State.Shooting) {
			return;
		}

		// skip forward to next step
		const waitTime = this.configuration.waitTimeInSeconds;
		const firstGroup = this.isFirstGroupOfEnd();
		const shootingEnd = firstGroup
			? this.addToDate(this.endStart, waitTime + this.getShootingTimeInSeconds())
			: this.addToDate(this.endStart, waitTime * 2 + this.getShootingTimeInSeconds() * 2);

		const remainingWaitingTime = Math.ceil(shootingEnd.getTime() - now.getTime());

		if (remainingWaitingTime > 0) {
			// cheat, and set endStart backwards by the difference we calculated
			this.endStart = new Date(this.endStart.valueOf() - remainingWaitingTime);
		}
	}

	public reset(): void {
		this.endCurrentEnd();
		this.initializeState();
	}

	// Should be called at least a few times per second, ideally in 10ths of seconds or faster
	public tick(now?: Date): ITournamentState {
		now = now || new Date();

		switch (this.state) {
			case State.Idle: return this.handleIdleState(now);
			case State.Waiting: return this.handleWaitingState(now);
			case State.Shooting: return this.handleShootingState(now);
			default: throw new Error(`Tournament is in invalid state: ${this.state}.`);
		}
	}

	private initializeState(): void {
		this.endStart = null;
		this.currentLight = 'red';
		this.currentShooters = 'A/B';
		this.endsShot = 0;
	}

	private handleIdleState(now: Date): ITournamentState {
		if (this.endStart === null) {
			// still idling
			return this.getCurrentState();
		} else if (now >= this.endStart) {
			// end has started, so change state:
			this.state = State.Waiting;
			// Waiting is start of preparation phase, so we need to honk 2 times
			return this.getCurrentState(this.configuration.waitTimeInSeconds, 2);
		}
	}

	private handleWaitingState(now: Date): ITournamentState {
		// we only need to check if waiting is over
		const waitingEnd = this.isFirstGroupOfEnd()
			? this.addToDate(this.endStart, this.configuration.waitTimeInSeconds)
			: this.addToDate(this.endStart, this.configuration.waitTimeInSeconds * 2 + this.getShootingTimeInSeconds());

		if (now <= waitingEnd) {
			// we're still waiting, so wait...
			const remainingWaitingTimeInFullSeconds = Math.ceil((waitingEnd.getTime() - now.getTime()) / 1000);
			return this.getCurrentState(remainingWaitingTimeInFullSeconds);
		} else {
			// waiting is over, so progress to next state
			this.state = State.Shooting;
			this.currentLight = 'green';
			return this.getCurrentState(this.getShootingTimeInSeconds(), 1);
		}
	}

	private handleShootingState(now: Date): ITournamentState {
		const waitTime = this.configuration.waitTimeInSeconds;
		const firstGroup = this.isFirstGroupOfEnd();
		const shootingEnd = firstGroup
			? this.addToDate(this.endStart, waitTime + this.getShootingTimeInSeconds())
			: this.addToDate(this.endStart, waitTime * 2 + this.getShootingTimeInSeconds() * 2);

		if (now <= shootingEnd) {
			const remainingTime = Math.ceil((shootingEnd.getTime() - now.getTime()) / 1000);

			// check if we need to switch to yellow phase
			if (this.currentLight === 'green' && (remainingTime <= this.configuration.yellowPhaseInSeconds)) {
				this.currentLight = 'yellow';
			}

			// we're still shooting, so continue counting
			return this.getCurrentState(remainingTime);
		} else {
			// shooting is over, so check whether we need switch shooters or stop the end
			if (this.configuration.alternatingShooters && firstGroup) {
				// do 2nd round
				this.state = State.Waiting;
				this.switchShooters();
				this.currentLight = 'red';
				return this.getCurrentState(this.configuration.waitTimeInSeconds, 2);
			} else {
				this.state = State.Idle;
				this.endStart = null;

				this.currentLight = 'red';
				this.endsShot += 1;

				if (this.endsShot >= this.configuration.numberOfEnds) {
					this.endsShot = 0;
				}

				return this.getCurrentState(null, 3);
			}
		}
	}

	private getCurrentState(secondsLeft?: number, honks?: Honks): ITournamentState {
		secondsLeft = Number.isInteger(secondsLeft) ? secondsLeft : null;
		honks = honks || 0;

		return {
			currentLight: this.currentLight,
			currentShooters: this.currentShooters,
			endsShot: this.endsShot,
			honks: honks,
			secondsLeft: secondsLeft,
			shooting: this.state !== State.Idle,
			totalEnds: this.configuration.numberOfEnds,
		};
	}

	private addToDate(date?: Date, offsetInSeconds?: number): Date {
		date = date || new Date();
		offsetInSeconds = offsetInSeconds || 0;
		const result = new Date(date.valueOf());
		result.setSeconds(result.getSeconds() + offsetInSeconds);
		return result;
	}

	private isFirstGroupOfEnd(): boolean {
		if (this.configuration.alternatingShooters) {
			const isOddEnd = (this.endsShot % 2 === 1);
			return isOddEnd
				? this.currentShooters === 'C/D'
				: this.currentShooters === 'A/B';
		} else {
			return true;
		}
	}

	private getShootingTimeInSeconds(): number {
		return this.configuration.secondsPerArrow * this.configuration.arrowsPerEnd;
	}

	private switchShooters(): void {
		this.currentShooters = (this.currentShooters === 'A/B')
			? 'C/D'
			: 'A/B';
	}
}

File: Models/Logic/Tournament.cs
using System;
using BscAmpel.Models;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

namespace BscAmpel.Models.Logic
{
	public class Tournament
	{
		private readonly ILogger _logger;
		private TournamentOptions _configuration;


		// Values for running the tournament
		private DateTime? _start;
		private Group _currentGroup;
		private int _endsShot;

		// helper variables, not actually necessary, but they do help a lot
		private RoundState _state = RoundState.Idle;
		private Light _light = Light.Red;

		public Tournament(ILogger<Tournament> logger, IOptions<TournamentOptions> tournamentOptions)
		{
			_logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
			_configuration = tournamentOptions?.Value ?? throw new System.ArgumentNullException(nameof(tournamentOptions));

			_logger.LogInformation("Tournament initialized with config {@Options}", _configuration);
			InitializeState();
		}

		public bool CanConfigure => (_start == null) && (_endsShot == 0);
		public void Configure(TournamentOptions newOptions)
		{
			_logger.LogDebug("{Method} has been called", nameof(Configure));

			if (newOptions == null) throw new ArgumentNullException(nameof(newOptions));
			if (!CanConfigure) throw new InvalidOperationException("Configuration can only be changed when idle.");

			_configuration = newOptions;
			InitializeState();
		}

		public void StartEnd(DateTime? now = null)
		{
			_logger.LogDebug("{Method} has been called", nameof(StartEnd));

			_start = now ?? DateTime.UtcNow;
		}

		public void EndCurrentEnd(DateTime? now = null)
		{
			_logger.LogDebug("{Method} has been called", nameof(EndCurrentEnd));

			if (_start == null)
				return;

			now = now ?? DateTime.UtcNow;

			var waitTime = _configuration.WaitTime;
			var shootingTime = _configuration.ShootingTime;

			var shootingEnd = IsFirstGroupdOfEnd()
				? _start.Value + waitTime + shootingTime
				: _start.Value + 2* waitTime + 2* shootingTime;

			var remainingWaitingTime = now - shootingEnd;
			if (remainingWaitingTime > TimeSpan.FromSeconds(1))
			{
				// cheat, and set back start date to what we calculated
				_start = _start.Value - remainingWaitingTime;
			}
		}

		public void Reset()
		{
			_logger.LogDebug("{Method} has been called", nameof(Reset));

			EndCurrentEnd();
			InitializeState();
		}

		// This is intended to be called *at least* a few times per second
		public TournamentState Tick(DateTime? now = null)
		{
			now = now ?? DateTime.UtcNow;

			switch (_state)
			{
				case RoundState.Idle: return HandleIdleState(now.Value);
				case RoundState.Waiting: return HandleWaitingState(now.Value);
				case RoundState.Shooting: return HandleShootingState(now.Value);
				default: throw new InvalidOperationException("Unknown state");
			}
		}

		private TournamentState HandleIdleState(DateTime now)
		{
			if (now >= _start)
			{
				_logger.LogInformation("Round start");

				// end has started, so change state
				_state = RoundState.Waiting;
				_logger.LogInformation("State changed from {StartState} to {TargetState}", RoundState.Idle, RoundState.Waiting);

				// Waiting is start of preparation phase, so we need to honk 2 times
				return GetCurrentState(_configuration.WaitTime, 2);
			}

			// else we're still idling
			return GetCurrentState();
		}

		// we only need to check if waiting is over
		private TournamentState HandleWaitingState(DateTime now)
		{
			var waitingEnd = IsFirstGroupdOfEnd()
				? _start.Value + _configuration.WaitTime
				: _start.Value + 2* _configuration.WaitTime + _configuration.ShootingTime;

			if (now <= waitingEnd)
			{
				// we're still waiting, so wait
				return GetCurrentState(waitingEnd - now);
			}
			else
			{
				// waiting is over, so start the shooting
				_state = RoundState.Shooting;
				_logger.LogInformation("State changed from {StartState} to {TargetState}", RoundState.Waiting, RoundState.Shooting);

				_light = Light.Green;
				_logger.LogInformation("Light changed from {StartLight} to {TargetLight}", Light.Red, Light.Green);

				// we started shooting: Honk once
				return GetCurrentState(_configuration.ShootingTime, 1);
			}
		}

		private TournamentState HandleShootingState(DateTime now)
		{
			var waitTime = _configuration.WaitTime;
			var isFirstGroup = IsFirstGroupdOfEnd();

			var shootingEnd = isFirstGroup
				? _start.Value + _configuration.WaitTime + _configuration.ShootingTime
				: _start.Value + 2* _configuration.WaitTime + 2* _configuration.ShootingTime;

			if (now <= shootingEnd)
			{
				// we're still shooting

				// check if we need to switch to yellow
				var remainingTime = shootingEnd - now;
				if ((_light != Light.Yellow) && (remainingTime <= _configuration.YellowPhase))
				{
					_logger.LogInformation("Entering yellow phase");
					_light = Light.Yellow;
					_logger.LogInformation("Light changed from {StartLight} to {TargetLight}", Light.Green, Light.Yellow);
				}

				return GetCurrentState(remainingTime);
			}
			else
			{
				_logger.LogInformation("Shooting time is over");

				// shooting is over. Check if we need to switch groups and continue or stop the end
				if (_configuration.AlternatingShooters && isFirstGroup)
				{
					_logger.LogInformation("Was first group and have more, so go to other group");

					// start 2nd round, honk twice
					_state = RoundState.Waiting;
					_logger.LogInformation("State changed from {StartState} to {TargetState}", RoundState.Shooting, RoundState.Waiting);

					SwitchGroups();

					_light = Light.Red;
					_logger.LogInformation("Light changed from {StartLight} to {TargetLight}", Light.Yellow, Light.Red);

					return GetCurrentState(_configuration.WaitTime, 2);
				}
				else
				{
					_logger.LogInformation("Was second group or only have one, so end this end.");

					// either, we only have one group, or we finished for second group of shooters, so end round
					_state = RoundState.Idle;
					_logger.LogInformation("State changed from {StartState} to {TargetState}", RoundState.Shooting, RoundState.Idle);

					_start = null;

					_light = Light.Red;
					_logger.LogInformation("Light changed from {StartLight} to {TargetLight}", Light.Yellow, Light.Red);

					_endsShot += 1;
					_logger.LogInformation("Number of ends shot: {NumberOfEnds}", _endsShot);


					if (_endsShot >= _configuration.NumberOfEnds)
					{
						_endsShot = 0;
						_logger.LogInformation("Tournament finished. Resetting number of ends shot: {NumberOfEnds}", _endsShot);
					}

					// End round = 3 honks
					return GetCurrentState(null, 3);
				}
			}
		}

		private bool IsFirstGroupdOfEnd()
		{
			if (!_configuration.AlternatingShooters)
				return true;

			// We record only complete ends shot. So _endsShot is BEHIND by 1 (i.e. still 0 while shooting the FIRST end)
			// So, if we have an even number in _endsShot (0, 2, 4...) we're starting with A/B shooters,
			// if we have an odd number (1, 3, 5...) this round starts with the C/D shooter group

			return ((_endsShot % 2 == 0)) // this tests for an even number
				? _currentGroup == Group.First
				: _currentGroup == Group.Second;
		}

		private void SwitchGroups()
		{
			if (_currentGroup == Group.First)
			{
				_currentGroup = Group.Second;
				_logger.LogInformation("Group changed from {StartGroup} to {TargetGroup}", Group.First, Group.Second);
			}
			else
			{
				_currentGroup = Group.First;
				_logger.LogInformation("Group changed from {StartGroup} to {TargetGroup}", Group.Second, Group.First);
			}
		}

		private TournamentState GetCurrentState(TimeSpan? timeLeft = null, int? honks = null)
		{
			return new TournamentState()
			{
				CurrentLight = _light,
				FirstGroup = _currentGroup == Group.First,
				EndsShot = _endsShot,
				TotalEnds = _configuration.NumberOfEnds,
				TimeLeft = timeLeft,
				Honks = honks ?? 0,
			};
		}

		private void InitializeState()
		{
			_state = RoundState.Idle;
			_light = Light.Red;
			_start = null;
			_currentGroup = Group.First;
			_endsShot = 0;
		}
	}
}
